### PR TITLE
feat(alert): add alert component

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
This is more or less an equivalent of our angular-ui/bootstrap implementation and I expect this proposal to trigger some discussion so... let it begin :-)

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
So, there are couple of usability issues with this component, as we've learned from angular-ui/bootstrap. The main issue is that the "close" button doesn't do anything useful by default (that is - it doesn't actually close alert). 

Unfortunately with the current Angular2 version there is no way of expressing "hey, component, go and destroy yourself" so the only way of proceeding would be to wrap everything in a `ngIf`:

```
@Component({
  selector: 'ngb-alert',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <template [ngIf]="dismissed">
    <div [class]="'alert alert-' + (type || 'warning')" role="alert">
      <template [ngIf]="dismissible">
        <button type="button" class="close" aria-label="Close" (click)="closeHandler()">
              <span aria-hidden="true">&times;</span>
        </button>
      </template>
      <ng-content></ng-content>
    </div>
    </template>
    `,
  directives: [NgIf]
})
export class NgbAlert {
  dismissed = false;
  @Input() dismissible = true;
  @Input() type;
  @Output() close = new EventEmitter();

  closeHandler() { dismissed = true; this.close.emit(null); }
}
```

but the downside of this is that:
* people can't capture / prevent closing of alerts
* the actual `<ngb-alert>` element would stay in the DOM after "closing" and this might screw up peoples' CSS styles

**Comment by pkozlowski-opensource** [comment]
Idea of an alternative API:

```html
<template ngb-alert (close)="onAlertClose($event)">
   Alert's content
</template>
```
and in `onClose` people could call `$event.preventDefault()` to prevent an alert from closing.

By having alerts on `<template>` elements we could easily "destroy" them without leaving any leftowers in the DOM.

Thoughts?

**Comment by wesleycho** [comment]
I don't quite like the original alert design in UI Bootstrap - it's a bit awkward.

Linking for reference the [Bootstrap 4 version](http://v4-alpha.getbootstrap.com/components/alerts/).

CSS-wise, this is a pretty straightforward component, so there is no reason to necessarily care about the pure CSS version. This leaves the dismissable version - there are two use cases for these sort of messages. There is the toastr type messages that pop on top of the whole UI, and there is the type that just show up in the container directly. These use cases would be best captured by utilizing the DCL I think, which also gives the ability to remove them & dynamically manage them.

WDYT about this approach?

**Comment by Foxandxss** [comment]
Alert and toasts perhaps have different goals, but worth a new discussion imo.

**Comment by pkozlowski-opensource** [comment]
So, my [previous idea](https://github.com/ng-bootstrap/core/pull/154#issuecomment-202896646) doesn't sound like an implementable one :-/ (or at least I don't know how to implement it...).

Regarding @wesleycho remarks:

 > I don't quite like the original alert design in UI Bootstrap - it's a bit awkward

I think we all agree here :-) The goal is to come up with sth better and this is what we are struggling with here....

> CSS-wise, this is a pretty straightforward component, so there is no reason to necessarily care about the pure CSS version.

Well, I still prefer to write `<ngb-alert type="warning">Watch out!</ngb-alert>`  as compared to dumping the whole HTML + CSS, remembering to add all the aria stuff, making sure that animations work etc. So still there is some value in it.

> This leaves the dismissable version

Yes, this is the crux. And I wouldn't drag toastr into the conversation - IMO toastr could build on top of "dumb" alerts.

**Comment by pkozlowski-opensource** [comment]
To sum up: as much as I don't like the design of the current ui-bootstrap alert it is only one I know how to implement :-/ To cover the admissible part we need one of:
* changes in the framework ("hey component, destroy yourself" or make [other API](https://github.com/ng-bootstrap/core/pull/154#issuecomment-202896646) implementable)
* new idea
* someone convincing that says that we are trying to solve an issue that doesn't exist

Otherwise we are stuck with the ui-bootstrap API / impl :-(

**Comment by pkozlowski-opensource** [comment]
Soooo, I had a chat with Misko yesterday on it (based on this doc: https://docs.google.com/document/d/1tMEhoAHhi-BZgHlbOMdxa5fu-Ev9PVM78SgC5uiGySk/edit?usp=sharing).

The tl;dr; is that:
- there will be mechanism to do `<template ngb-alert>` (it is not implementable yet but you can have a look at the gist toward the end of the linked doc);
- we still going to need a "wrapper" component as in this PR.

**Comment by pkozlowski-opensource** [comment]
My suggestion based on the above: land what we've got in this PR and improve close support as soon as we've got sufficient mechanism in the framework. @Foxandxss @wesleycho WDYT?

**Comment by wesleycho** [comment]
I'm fine with that.

**Comment by Foxandxss** [comment]
You were the people who didn't want to add some features that needed to be changed later. I am fine with the idea tho.

**Comment by wesleycho** [comment]
Well, we need to move faster (and I need to spend more time with ng-bootstrap), so I can bend.

**Comment by pkozlowski-opensource** [comment]
@Foxandxss for the record - we will still need the exact implementation from this PR even if support in Angular gets better. So no code will be removed afterwards.

**Comment by Foxandxss** [comment]
Yeah, you are right.

**Comment by pkozlowski-opensource** [comment]
Any other comments or is this one good to go?

File: src/alert/alert.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import {Component, Input, Output, EventEmitter, ChangeDetectionStrategy} from 'angular2/core';
+import {NgIf} from 'angular2/common';
```
Comment (Foxandxss) [inline src/alert/alert.ts]: out

File: src/alert/alert.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    <div [class]="'alert alert-' + (type || 'warning')" role="alert">
+      <template [ngIf]="dismissible">
+        <button type="button" class="close" aria-label="Close" (click)="closeHandler()">
+              <span aria-hidden="true">&times;</span>
+        </button>
+      </template>
+      <ng-content></ng-content>
+    </div>
+    `,
+  directives: [NgIf]
```
Comment (Foxandxss) [inline src/alert/alert.ts]: out

File: src/alert/alert.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+})
+export class NgbAlert {
+  @Input() dismissible = true;
+  @Input() type;
+  @Output() close = new EventEmitter();
+
+  closeHandler() { this.close.emit(null); }
```
Comment (Foxandxss) [inline src/alert/alert.ts]: is null needed?

File: src/alert/alert.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        </button>
+      </template>
+      <ng-content></ng-content>
+    </div>
+    `,
+  directives: [NgIf]
+})
+export class NgbAlert {
+  @Input() dismissible = true;
+  @Input() type;
```
Comment (Foxandxss) [inline src/alert/alert.ts]: could we have type in here?

**Comment by Foxandxss** [comment]
Apart from what I said, perhaps some extra tests, but that can be done later (I want to revisit tests and do a few)

**Pull request closed without merging**